File: remocra/src/main/java/fr/sdis83/remocra/domain/remocra/Rci.java
package fr.sdis83.remocra.domain.remocra;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.validation.constraints.NotNull;

import org.hibernate.annotations.Formula;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.annotations.Type;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.roo.addon.javabean.RooJavaBean;
import org.springframework.roo.addon.jpa.activerecord.RooJpaActiveRecord;
import org.springframework.roo.addon.json.RooJson;
import org.springframework.roo.addon.tostring.RooToString;

import com.vividsolutions.jts.geom.Geometry;
import com.vividsolutions.jts.geom.Point;

import flexjson.JSONDeserializer;
import flexjson.JSONSerializer;
import fr.sdis83.remocra.domain.utils.RemocraDateHourFormat;
import fr.sdis83.remocra.domain.utils.RemocraDateHourTransformer;
import fr.sdis83.remocra.util.Featurable;
import fr.sdis83.remocra.util.Feature;
import fr.sdis83.remocra.web.deserialize.GeometryFactory;

@RooJavaBean
@RooToString
@RooJson
@RooJpaActiveRecord(/* versionField = "version" */)
public class Rci implements Featurable {

    public static enum DirectionVent {
        N, S, E, O, NE, NO, SE, SO
    }

    static JSONSerializer SERIALIZER = new JSONSerializer();

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    // @Version
    // @Column(name = "version", columnDefinition = "INTEGER default 1")
    // private Integer version;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date dateModification;

    @NotNull
    @ManyToOne
    private Utilisateur utilisateur;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date dateIncendie;

    @NotNull
    @ManyToOne
    TypeRciOrigineAlerte origineAlerte;

    @ManyToOne
    private Commune commune;

    private String voie;

    private String complement;

    @NotNull
    @Type(type = "org.hibernate.spatial.GeometryType")
    private Point geometrie;

    @Formula("ST_AsGeoJSON(geometrie)")
    private String jsonGeometrie;

    private String coordDFCI;

    @ManyToOne
    private Utilisateur arriveeDdtmOnf;

    @ManyToOne
    private Utilisateur arriveeSdis;

    @ManyToOne
    private Utilisateur arriveeGendarmerie;

    @ManyToOne
    private Utilisateur arriveePolice;

    @NotNull
    private String pointEclosion;

    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date gdh;

    private Boolean ventLocal;

    private Integer hygrometrie;

    private DirectionVent directionVent;

    private Double temperature;

    /**
     * 0 à 12
     */
    private Integer forceVent;

    /**
     * 1 à 11
     */
    private Integer indiceRothermel;

    private Double superficieSecours;
    private Double superficieReferent;
    private Double superficieFinale;

    private String premierEngin;

    private String premierCos;

    private String forcesOrdre;

    private Boolean gelLieux;

    @ManyToOne
    private TypeRciPromCategorie categoriePromethee;

    @ManyToOne
    private TypeRciDegreCertitude degreCertitude;

    private String commentaireConclusions;

    @OnDelete(action = OnDeleteAction.CASCADE)
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "rci", orphanRemoval = true)
    private Set<RciDocument> rciDocuments = new HashSet<RciDocument>();

    public static Rci fromJsonToRci(String json) {
        return new JSONDeserializer<Rci>().use(null, Rci.class).use(Date.class, RemocraDateHourTransformer.getInstance()).use(Geometry.class, new GeometryFactory())
                .deserialize(json);
    }

    public Feature toFeature() {
        Feature feature = new Feature(this.id, this.getJsonGeometrie());
        feature.addProperty("dateIncendie", new RemocraDateHourFormat().format(getDateIncendie()));
        feature.addProperty("coordDFCI", this.getCoordDFCI());
        return feature;
    }
}

File: remocra/src/main/webapp/js/app/remocra/model/Rci.js
Ext.require('Sdis.Remocra.network.RemocraRest');
Ext.require('Sdis.Remocra.model.TypeRciDegreCertitude');
Ext.require('Sdis.Remocra.model.TypeRciOrigineAlerte');
Ext.require('Sdis.Remocra.model.TypeRciRisqueMeteo');
Ext.require('Sdis.Remocra.model.TypeRciPromCategorie');
Ext.require('Sdis.Remocra.model.RciDocument');

Ext.define('Sdis.Remocra.model.Rci', {
    extend : 'Ext.data.Model',

    fields : [ {
        name : 'id',
        type : 'int',
        useNull : true
    }, {
        name : 'dateModification',
        type : 'date',
        dateFormat : 'c'
    }, {
        name : 'dateIncendie',
        type : 'date',
        dateFormat : 'c'
    }, {
        name : 'voie',
        type : 'string',
        useNull : true
    }, {
        name : 'complement',
        type : 'string',
        useNull : true
    }, {
        name : 'geometrie',
        type : 'string'
    }, {
        name : 'jsonGeometrie',
        type : 'auto',
        persist : false,
        useNull : true
    }, {
        name : 'coordDFCI',
        type : 'string'
    }, {
        name : 'pointEclosion',
        type : 'string',
        useNull : true
    }, {
        name : 'gdh',
        type : 'date',
        dateFormat : 'c'
    }, {
        name : 'ventLocal',
        type : 'boolean',
        convert: function (v, record){ // v = nouvelle valeur, sert a convertir un int en boolean
            if(v == 1){
                return true;
            } else if(v == 0){
                   return false;
               } else if(v == -1 || v == null){
                    return null;
               }
        },
        useNull : true,
        allowNull : true
    }, {
        name : 'hygrometrie',
        type : 'int',
        useNull : true
    }, {
        name : 'directionVent',
        type : 'string',
        convert: function (v, record){
            if (v == 'none'){
                return null;
            }
            return v;
        },
        useNull : true
    }, {
        name : 'temperature',
        type : 'float',
        useNull : true
    }, {
        name : 'forceVent',
        type : 'int',
        useNull : true
    }, {
        name : 'indiceRothermel',
        type : 'int',
        useNull : true
    }, {
        name : 'superficieSecours',
        type : 'float',
        useNull : true
    }, {
        name : 'superficieReferent',
        type : 'float',
        useNull : true
    }, {
        name : 'superficieFinale',
        type : 'float',
        useNull : true
    }, {
        name : 'premierEngin',
        type : 'string',
        useNull : true
    }, {
        name : 'premierCos',
        type : 'string',
        useNull : true
    }, {
        name : 'forcesOrdre',
        type : 'string',
        useNull : true
    }, {
        name : 'gelLieux',
        type : 'boolean',
        convert: function (v, record){ // v = nouvelle valeur, sert a convertir un int en boolean
            if(v == 1){
                return true;
            } else if(v == 0){
                   return false;
               } else if(v == -1 || v == null){
                    return null;
               }
        },
        useNull : true,
        allowNull : true
    }, {
        name : 'commentaireConclusions',
        type : 'string',
        useNull : true
    }

    ],

    associations : [ {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'utilisateur',
        getterName : 'getUtilisateur',
        setterName : 'setUtilisateur',
        associatedName : 'Utilisateur',
        persist : false
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'arriveeDdtmOnf',
        getterName : 'getArriveeDdtmOnf',
        setterName : 'setArriveeDdtmOnf',
        associatedName : 'ArriveeDdtmOnf',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'arriveeGendarmerie',
        getterName : 'getArriveeGendarmerie',
        setterName : 'setArriveeGendarmerie',
        associatedName : 'ArriveeGendarmerie',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'arriveePolice',
        getterName : 'getArriveePolice',
        setterName : 'setArriveePolice',
        associatedName : 'ArriveePolice',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'arriveeSdis',
        getterName : 'getArriveeSdis',
        setterName : 'setArriveeSdis',
        associatedName : 'ArriveeSdis',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciOrigineAlerte',
        associationKey : 'origineAlerte',
        getterName : 'getOrigineAlerte',
        setterName : 'setOrigineAlerte',
        associatedName : 'OrigineAlerte',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Commune',
        associationKey : 'commune',
        getterName : 'getCommune',
        setterName : 'setCommune',
        associatedName : 'Commune',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciPromCategorie',
        associationKey : 'categoriePromethee',
        getterName : 'getCategoriePromethee',
        setterName : 'setCategoriePromethee',
        associatedName : 'CategoriePromethee',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciPromPartition',
        associationKey : 'partitionPromethee',
        getterName : 'getPartitionPromethee',
        setterName : 'setPartitionPromethee',
        associatedName : 'PartitionPromethee',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciPromFamille',
        associationKey : 'famillePromethee',
        getterName : 'getFamillePromethee',
        setterName : 'setFamillePromethee',
        associatedName : 'FamillePromethee',
        persist : true
    },{
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciRisqueMeteo',
        associationKey : 'risqueMeteo',
        getterName : 'getRisqueMeteo',
        setterName : 'setRisqueMeteo',
        associatedName : 'RisqueMeteo',
        persist : true
    },{
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeRciDegreCertitude',
        associationKey : 'degreCertitude',
        getterName : 'getDegreCertitude',
        setterName : 'setDegreCertitude',
        associatedName : 'DegreCertitude',
        persist : true
    }, {
        type : 'hasMany',
        model : 'Sdis.Remocra.model.RciDocument',
        associationKey : 'rciDocuments',
        name : 'rciDocuments',
        persist : false
    } ],

    proxy : {
        type : 'remocra.rest',
        url : Sdis.Remocra.util.Util.withBaseUrl('../rci')
    }
});